<a href="https://colab.research.google.com/github/jostina/Week-1-2-Data-Cleaning-Exploratory-Data-Analysis/blob/main/Week_4_Supervised_Learning_Titanic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Import libraries
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.impute import SimpleImputer
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [2]:
# Load Titanic Dataset
df = pd.read_csv("/content/titanic_cleaned.csv")


In [4]:
print("First Five Rows")
df.head()

First Five Rows


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,HasCabin,FamilySize,IsAlone,Title,AgeGroup
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,0,2,0,Mr,Young Adult
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,1,2,0,Mrs,Adult
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,0,1,1,Miss,Young Adult
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,1,2,0,Mrs,Young Adult
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,0,1,1,Mr,Young Adult


In [5]:
print("\nDataset Information")
df.info()


Dataset Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    object 
 11  HasCabin     891 non-null    int64  
 12  FamilySize   891 non-null    int64  
 13  IsAlone      891 non-null    int64  
 14  Title        891 non-null    object 
 15  AgeGroup     891 non-null    object 
dtypes: float64(2), int64(8), object(6)
memory usage: 111.5+ KB


# Data Preprocessing

In [6]:
# Drop columns that are not useful for prediction
columns_to_drop = ["PassengerId", "Name", "Ticket", "Cabin"]

for col in columns_to_drop:
    if col in df.columns:
        df.drop(col, axis=1, inplace=True)

In [8]:
# Convert categorical variables into numeric values
df = pd.get_dummies(df, drop_first=True)

print("\nCleaned Dataset")
df.head()


Cleaned Dataset


,Survived,Pclass,Age,SibSp,Parch,Fare,HasCabin,FamilySize,IsAlone,Sex_male,Embarked_Q,Embarked_S,Title_Miss,Title_Mr,Title_Mrs,Title_Rare,AgeGroup_Child,AgeGroup_Senior,AgeGroup_Teen,AgeGroup_Young Adult
0,0,3,22.0,1,0,7.2500,0,2,0,True,False,True,False,True,False,False,False,False,False,True
1,1,1,38.0,1,0,71.2833,1,2,0,False,False,False,False,False,True,False,False,False,False,False
2,1,3,26.0,0,0,7.9250,0,1,1,False,False,True,True,False,False,False,False,False,False,True
3,1,1,35.0,1,0,53.1000,1,2,0,False,False,True,False,False,True,False,False,False,False,True
4,0,3,35.0,0,0,8.0500,0,1,1,True,False,True,False,True,False,False,False,False,False,True


In [9]:
# Features and Target
X = df.drop("Survived", axis=1)
y = df["Survived"]


In [10]:
# Train/Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print("\nTraining Samples:", len(X_train))
print("Testing Samples:", len(X_test))


Training Samples: 712
Testing Samples: 179


In [11]:
# Train Logistic Regression Model
model = LogisticRegression(max_iter=1000)

model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [12]:
# Make Predictions
predictions = model.predict(X_test)

In [13]:
# Evaluate Model
accuracy = accuracy_score(y_test, predictions)

print("\nAccuracy Score")
print(round(accuracy, 4))

print("\nConfusion Matrix")
print(confusion_matrix(y_test, predictions))

print("\nClassification Report")
print(classification_report(y_test, predictions))


Accuracy Score
0.8156

Confusion Matrix
[[89 16]
 [17 57]]

Classification Report
              precision    recall  f1-score   support

           0       0.84      0.85      0.84       105
           1       0.78      0.77      0.78        74

    accuracy                           0.82       179
   macro avg       0.81      0.81      0.81       179
weighted avg       0.82      0.82      0.82       179



# Brief Explanation of the Results

The Logistic Regression model predicts whether a passenger survived based on features such as age, sex, passenger class, fare, and embarkation port. The model's performance is measured using the Accuracy Score, which indicates the proportion of correct predictions. For example, an accuracy of 82% means the model correctly classified about 82 out of every 100 passengers. A higher accuracy generally indicates better predictive performance, though the confusion matrix and classification report provide additional insight into how well the model identifies survivors and non-survivors.